### ISSUE TITLE
`onchain_feerate` is `high fees` when only waiting for two blocks?

### ISSUE THREAD

**Issue by GhostTyper** [ZmnSCPxj/clboss#88]
Even though we actually have low fees (1 sat/vbyte, less than two blocks in the mempool) clboss considers them as high and is not opening new channels:

![image](https://user-images.githubusercontent.com/22769697/145453969-3d8fb018-3265-45a3-a5ad-022c951e83b3.png)

Result of `lightning-cli clboss-status`:

    "onchain_feerate": {
       "hi_to_lo": 253,
       "init_mid": 253,
       "lo_to_hi": 253,
       "last_feerate_perkw": 400,
       "judgment": "high fees"
    },

**Comment by ZmnSCPxj** [comment]
CLBOSS judges feerate as "low" based on historical data for the past two weeks.. The `hi_to_lo` setting being 253 means it is waiting for feerates to drop to 253sat/kweight (about 1.0sat/vbyte).  However, the current feerate is 400sat/kweight (about 1.58sat/vbyte).

The `hi_to_lo` setting being 253 means that for the past two weeks, CLBOSS has consistently seen feerates as low as 1sat/vbyte as being confirmed on the network over the past two weeks, and the current 1.58sat/vbyte feerates are 58% more expensive than that.  CLBOSS is hoping that the feerates will drop back down to 1.0 sat/vbyte.

Basically --- CLBOSS does not round off to whole sat/vbyte the way most explorers show, and the 1.58sat/vbyte is making it pause because it saw that 1.0sat/vbyte was working before and it is hoping that the 1.0sat/vbyte will be seen again.

**Comment by ZmnSCPxj** [comment]
Other policies could be added --- would you consider, say, that all *absolute* feerates below, say, 2.0sat/vbyte are always "low" regardless of historical data showing that 1.0sat/vbytes were getting confirmed on the network?

**Comment by mrsieb** [comment]
Interesting, i have zero experience with clbos but rethinking how I personally rate and predigt the mempool, I simply look at the possibility  if the mempool can get cleared out in a at least a couple of days.  So if 1.0sat/vbyte was working the past weeks and there a just few blocks in the mempool, taking also in account the ‎ incoming vB/s, there is no reason to assume the 1.0sat/vbyte  gets stuck.

**Comment by GhostTyper** [comment]
> would you consider, say, that all _absolute_ feerates below, say, 2.0sat/vbyte are always "low" regardless of historical data showing that 1.0sat/vbytes were getting confirmed on the network?

I would say something like that. I think that all transactions are done with Replace-By-Fee possibility? If so, we could in a low amount of blocks scenario or where we are very near to 1 sat/vbyte within the next 3 blocks just use 1 sat/vbyte?

**Comment by ZmnSCPxj** [comment]
>  I think that all transactions are done with Replace-By-Fee possibility?

Yes, but note that C-lightning itself has not been well-tested to handle change outputs in an RBF situation; i.e. C-Lightning will insert change outputs in its db immediately as soon as you build the transaction, but if you later RBF the same transaction, those outputs will change, but C-Lightning will not update its database (and in particular, it is possible to build later transactions spending the change outputs, but if the first transaction is RBFed then the later transactions are not rebuilt and C-Lightning could lose track of funds).  RBF is *complicated*, this is why uptake of RBF in onchain wallets was classically slow, and new onchain wallets (like the onchain wallet of C-Lightning) often take a while before they can really reliably use RBF without losing funds.  Note in particular that CLBOSS will never use funds that have not been confirmed for at least 3 blocks (or was it 6...? something, anyway) just to avoid RBF-related changing of txouts, so *maybe* we can RBF `fundchannel` transactions.... maybe, ***if*** C-Lightning had an interface to do that (which it does not, it seems --- `openchannel_bump` requires the caller to do a fair amount more work than just bumping, the caller needs to keep track of the vouts spent and so on.... and there is always the race condition that the current transaction gets confirmed while you are building the next, RBFed transaction, and I am uncertain if the `openchannel_*` mechanism handles that edge case well).

(Things might have changed since I last checked this a year ago; in particular the new "Open Channel V2" mechanism might defer change outputs being put into the db until confirmation of the transaction, and "Open Channel V2" in particular integrates RBF (whereas RBF channel opens pre-"Open Channel V2" are arguably hacks, that cannot be done by C-Lightning due to its "one channel per peer" rule), but I believe "Open Channel V2" is gated behind an `experimental` flag; so we cannot rely on RBF on CLBOSS layer, we always use the "best" estimate the C-Lightning has for feerates and never RBF.)

Worse, CLBOSS is automation, and that means the human operating the node with CLBOSS might very well not keep track of what CLBOSS is doing after some time.  Thus having CLBOSS perform RBF is risky unless we are absolutely sure that the underlying C-Lightning handles race conditions correctly (in particular, we broadcast a tx, then CLBOSS decides to feebump, then while getting signatures and etc the current tx gets confirmed, does C-Lightning handle that correctly and use the previous channel ID instead of the feebumped one? --- There Ain't No Such Thing As A Global Mempool, and just because our own mempool has the feebumped tx does ***not*** mean the next lucky miner mempool has the feebumped one, it could be the original tx!), it is far safer to just use the confirm-on-next-block feerate and never RBF and to defer spending even at 1.58 sat/vbyte (because recently 1.0sat/vbyte was good) rather than risk funds loss due to bugs in RBF handling, because ***RBF IS HARD***.

**Comment by ZmnSCPxj** [comment]
Okay, so C-Lightning has a `test_rbf_non_last_mined` that tests the situation where a non-last RBF is what gets mined, which covers that case at least.  I have to go look at change handling, and also see why the test has a `@pytest.mark.openchannel` tag, which seems to imply that the test only runs if openchannel is compiled in, which implies that not all configurations have it.

Even with this, we should note that the API available is low-level and inconvenient; we should probably modify C-Lightning first to provide a convenient feebump API for channel funding.

**Comment by GhostTyper** [comment]
But also if we don't change anything there is a chance that we can get into the situation where we would need to RBF, because we won't be mined else. It's just that the chance is very low, right?

**Comment by ZmnSCPxj** [comment]
> because we won't be mined else

Funds are reserved for 72 blocks by `fundpbt` and friends, so you can replace the low-feerate with another transaction later when they are unreserved, at the now higher feerate.  It probably means you have some unconfirmed channels with peers, but the peer will forget those after 2016 blocks (fundee forgets, but funder remembers).

It is also a reason to always use a highballed initial feerate.  CLBOSS uses `normal` feerate, which "should confirm" within 3 blocks or so, supposedly, well below the 72 block reservation.

In any case, if we build only one transaction, we have the assurance that only that transaction will get confirmed, and in particular, any subsequent transaction that spend change outputs will not get invalidated out from under us due to replacements occurring.  So there is strictly fewer places to lose funds in the conservative "highball the feerate" model.

**Comment by ZmnSCPxj** [comment]
See https://github.com/ElementsProject/lightning/issues/4971 as well.  It needs to be implemented and tested at C-Lightning first before we can rely on feebumping on the CLBOSS layer.

**Comment by ZmnSCPxj** [comment]
RBFing funding txes should probably go to a new issue, I think?  This should be about feerate-for-within-3-blocks that are, in an absolute sense, "low", being judged as "high" because historically, even lower feerates were seen.  RBF is an issue when the feerates are absolutely high, anyway.
